### PR TITLE
Prevent 500 errors for attachments with long filenames

### PR DESCRIPTION
**Pull request by ckumar1**
Truncate long filenames when creating tempfiles
 
/cc @zendesk/sustaining

### References
 - https://support.zendesk.com/agent/tickets/1298196

### Risks
 - None

### CODE REVIEW

File: lib/technoweenie/attachment_fu.rb [hunk 1]
```diff
@@ -517,7 +517,11 @@ def logger
 
         # Generates a unique filename for a Tempfile.
         def random_tempfile_filename
-          "#{rand Time.now.to_i}#{filename || 'attachment'}"
+          if filename.length > 215
+            "#{rand Time.now.to_i}#{filename.truncate(235, omission: '') || 'attachment'}"
```
Comment (miketineo) [inline lib/technoweenie/attachment_fu.rb]: I'd say you'll never use ``'attachment'`` here provided ``filename.length > 215`` is true

File: lib/technoweenie/attachment_fu.rb [hunk 2]
```diff
@@ -517,7 +517,11 @@ def logger
...
+          else
+            "#{rand Time.now.to_i}#{filename || 'attachment'}"
+          end
         end
```
Comment (miketineo) [inline lib/technoweenie/attachment_fu.rb]: you might use
```
temp_filename = filename.length > 215 ? filename.truncate(235, omission: '') : filename 
"#{rand Time.now.to_i}#{temp_filename || 'attachment'}"
```

**Comment by miketineo** [comment]
:+1:

**Comment by pdeuter** [comment]
Do we know why the filenames are so long?  Is it due to utf-8 encoding issues?  (i.e. filenames with CJK characters encoded with 3 bytes?)  If yes, then these filenames probably should be allowed.  Would it be possible to accommodate longer filenames by making the column wider?  (Note: we would still need code to check the limit - whatever it ended up being - and truncate.)

File: lib/technoweenie/attachment_fu.rb (hunk 2 above)
Comment (ckumar1) [inline lib/technoweenie/attachment_fu.rb]: Yeah, I think I'll use this instead. It takes care of `'attachment'` being duplicated too

**Comment by ckumar1** [comment]
The attachment filenames were long because the end-user was testing to make sure we supported filenames up to 255 chars to match typical OS limits on filenames. The filenames were getting too long because of the random strings that the `random_tempfile_filename` method and the `make_tmpname` method add to the original filename.

**Comment by pdeuter** [comment]
+1

File: lib/technoweenie/attachment_fu.rb [hunk 3]
```diff
@@ -517,7 +517,8 @@ def logger
 
         # Generates a unique filename for a Tempfile.
         def random_tempfile_filename
-          "#{rand Time.now.to_i}#{filename || 'attachment'}"
+          temp_filename = filename.length > 215 ? filename.truncate(215, omission: '') : filename
```
Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: can always call truncate

Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: or just `filename.slice(0, 215)`

File: lib/technoweenie/attachment_fu.rb [hunk 4]
```diff
@@ -517,7 +517,8 @@ def logger
 
         # Generates a unique filename for a Tempfile.
         def random_tempfile_filename
-          "#{rand Time.now.to_i}#{filename || 'attachment'}"
+          temp_filename = filename.length > 215 ? filename.truncate(215, omission: '') : filename 
+          "#{rand Time.now.to_i}#{temp_filename || 'attachment'}"
```
Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: `||` can go away since it's never nil

Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: no need to call Time.now ... just use a big number like 999999999

Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: `"#{rand 9999999}#{filename.slice(0, 215)}"`

Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: also missing a test

Comment (ckumar1) [inline lib/technoweenie/attachment_fu.rb]: Where should I put the test? This method doesn't appear to be getting tested yet.

Comment (ckumar1) [inline lib/technoweenie/attachment_fu.rb]: I added in the fixes in #17

Comment (grosser) [inline lib/technoweenie/attachment_fu.rb]: put a rais in it and see what blows up ... or just add a new test somewhere
:)

Comment (ckumar1) [inline lib/technoweenie/attachment_fu.rb]: I tried putting a raise in and there seems to be too much blowing up to be helpful. 
I'm going the route of adding a new tag, but I'm not sure how to access the `create_temp_file` or `write_to_temp_file` methods so I can test them.

**Comment by grosser** [comment]
fixes and tests are released as 1.0.5

**Pull request merged**